Software/GuitarPedal/Util/pitch_shifter.h

```diff
@@ -0,0 +1,187 @@
+                        ((transpose_ < 0.f) && prev_phs_a_ < fade1);
+    bool recalc_b_fun = ((transpose_ >= 0.f) && (prev_phs_b_ > fade2)) ||
+                        ((transpose_ < 0.f) && prev_phs_b_ < fade2);
+    // Randomly trigger mod recalc when tranpose is 0 so that the "fun"
+    // parameter still does something when not transposing
+    bool rand_retrig = (transpose_ >= -0.25f && transpose_ < 0.25f);
+    if (rand_retrig && fun_ > 0.f)
+      if (myrand() % 65536 < 4) recalc_a_fun = recalc_b_fun = true;
+    if (recalc_a_fun) {
+      mod_a_amt_ =
+          fun_ * ((float)(myrand() % 255) / 255.0f) * (del_size_ * 0.5f);
+      mod_coeff_[0] = 0.0002f + (((float)(myrand() % 255) / 255.0f) * 0.001f);
+    }
+    if (recalc_b_fun) {
+      mod_b_amt_ =
+          fun_ * ((float)(myrand() % 255) / 255.0f) * (del_size_ * 0.5f);
+      mod_coeff_[1] = 0.0002f + (((float)(myrand() % 255) / 255.0f) * 0.001f);
+    }
+    slewed_mod_[0] += mod_coeff_[0] * (mod_a_amt_ - slewed_mod_[0]);
+    slewed_mod_[1] += mod_coeff_[1] * (mod_b_amt_ - slewed_mod_[1]);
+    prev_phs_a_ = fade1;
+    prev_phs_b_ = fade2;
+    fade1 = 1.0f - fade1;
+    fade2 = 1.0f - fade2;
+    mod_[0] = fade1 * (del_size_ - 1);
+    mod_[1] = fade2 * (del_size_ - 1);
+#ifdef USE_ARM_DSP
+    gain_[0] = arm_sin_f32(fade1 * (float)M_PI);
+    gain_[1] = arm_sin_f32(fade2 * (float)M_PI);
+#else
+    gain_[0] = sinf(fade1 * PI_F);
+    gain_[1] = sinf(fade2 * PI_F);
+#endif
+
+    // Handle Delay Writing
+    d_[0].Write(in);
+    d_[1].Write(in);
+    // Modulate Delay Lines
+    // d_[0].SetDelay(mod_[0] + mod_a_amt_);
+    // d_[1].SetDelay(mod_[1] + mod_b_amt_);
+    d_[0].SetDelay(mod_[0] + slewed_mod_[0]);
+    d_[1].SetDelay(mod_[1] + slewed_mod_[1]);
+    val = 0.0f;
+    val += (d_[0].Read() * gain_[0]);
+    val += (d_[1].Read() * gain_[1]);
+    return val;
+  }
+
+  /** sets transposition in semitones
+   */
+  void SetTransposition(const float &transpose) {
+    float ratio;
+    if (transpose_ != transpose || force_recalc_) {
+      transpose_ = quantize_semitones_ ? (int32_t)transpose : transpose;
+      ratio = pow(2.f, transpose_ / 12.f);
+      mod_freq_ = ((ratio - 1.0f) * sr_) / del_size_;
+      phs_[0].SetFreq(mod_freq_);
+      phs_[1].SetFreq(mod_freq_);
+      if (force_recalc_) {
+        force_recalc_ = false;
+      }
+    }
+  }
+
+  /** sets delay size changing the timbre of the pitchshifting
+   */
+  void SetDelSize(uint32_t size) {
+    del_size_ = size < kShiftDelaySize ? size : kShiftDelaySize;
+    force_recalc_ = true;
+    SetTransposition(transpose_);
+  }
+
+  /** sets an amount of internal random modulation, kind of sounds like
+   * tape-flutter
+   */
+  inline void SetFun(float f) { fun_ = f; }
+
+ private:
+  /** Shift can be 30-100 ms lets just start with 50 for now.
+  0.050 * SR = 2400 samples (at 48kHz) */
+  // static constexpr size_t kShiftDelaySize = 16384;
+  static constexpr size_t kShiftDelaySize = 2048;
```

Comment:
setting this to ~5000+ makes a full octave sound pretty good, but the latency is pretty terrible and the buffers take up a lot of space. I could move stuff to SDRAM probably, but that won't fix the latency, and reallocating with a changing value here based on the selected octave seems not worth it really. Easiest would probably be to just remove the full octave setting for now, and just stay at +-5 or so semitones for now, but the others aren't hurting anything, they just don't sound good